src/visitor/mssql.rs

```diff
@@ -79,31 +79,65 @@ impl<'a> Visitor<'a> for Mssql<'a> {
     }
 
     fn visit_equals(&mut self, left: Expression<'a>, right: Expression<'a>) -> visitor::Result {
```

Comment:
This and the `visit_not_equals` can be written without cloning (which can be very expensive here):

```patch
diff --git a/src/ast/expression.rs b/src/ast/expression.rs
index c349a91..0a2b2cf 100644
--- a/src/ast/expression.rs
+++ b/src/ast/expression.rs
@@ -51,11 +51,7 @@ impl<'a> Expression<'a> {
 
     #[cfg(feature = "xml")]
     pub(crate) fn is_xml_value(&self) -> bool {
-        match &self.kind {
-            ExpressionKind::Parameterized(Value::Xml(_)) => true,
-            ExpressionKind::Value(expr) => expr.is_xml_value(),
-            _ => false,
-        }
+        self.kind.is_xml_value()
     }
 
     pub(crate) fn is_asterisk(&self) -> bool {
@@ -186,6 +182,17 @@ pub enum ExpressionKind<'a> {
     Value(Box<Expression<'a>>),
 }
 
+impl<'a> ExpressionKind<'a> {
+    #[cfg(feature = "xml")]
+    pub(crate) fn is_xml_value(&self) -> bool {
+        match self {
+            Self::Parameterized(Value::Xml(_)) => true,
+            Self::Value(expr) => expr.is_xml_value(),
+            _ => false,
+        }
+    }
+}
+
 /// A quick alias to create an asterisk to a table.
 pub fn asterisk() -> Expression<'static> {
     Expression {
diff --git a/src/visitor/mssql.rs b/src/visitor/mssql.rs
index a5b47d7..15f7b57 100644
--- a/src/visitor/mssql.rs
+++ b/src/visitor/mssql.rs
@@ -1,4 +1,5 @@
 use super::Visitor;
+use crate::prelude::Aliasable;
 use crate::prelude::Query;
 use crate::{
     ast::{Column, Expression, ExpressionKind, Insert, IntoRaw, Merge, OnConflict, Order, Ordering, Row, Values},
@@ -79,18 +80,32 @@ impl<'a> Visitor<'a> for Mssql<'a> {
     }
 
     fn visit_equals(&mut self, left: Expression<'a>, right: Expression<'a>) -> visitor::Result {
-        match (&left.kind, &right.kind) {
+        match (left.kind, right.kind) {
             // we can't compare with tuples, so we'll convert it to an AND
             (ExpressionKind::Row(left), ExpressionKind::Row(right)) => {
-                self.visit_multiple_tuple_comparison(left.clone(), Values::from(iter::once(right.clone())), false)?;
+                self.visit_multiple_tuple_comparison(left, Values::from(iter::once(right)), false)?;
             }
-            (_left_kind, _right_kind) => {
+            (left_kind, right_kind) => {
+                let (l_alias, r_alias) = (left.alias, right.alias);
+
                 #[cfg(feature = "xml")]
-                let left_xml = left.is_xml_value();
+                let (left_xml, right_xml) = (left_kind.is_xml_value(), right_kind.is_xml_value());
+
+                let mut left = Expression::from(left_kind);
+
+                if let Some(alias) = l_alias {
+                    left = left.alias(alias);
+                }
+
+                let mut right = Expression::from(right_kind);
+
+                if let Some(alias) = r_alias {
+                    right = right.alias(alias);
+                }
 
                 match self {
                     #[cfg(feature = "xml")]
-                    _ if right.is_xml_value() => {
+                    _ if right_xml => {
                         self.surround_with("CAST(", " AS NVARCHAR(MAX))", |x| x.visit_expression(left))?
                     }
                     _ => self.visit_expression(left)?,
@@ -112,18 +127,32 @@ impl<'a> Visitor<'a> for Mssql<'a> {
     }
 
     fn visit_not_equals(&mut self, left: Expression<'a>, right: Expression<'a>) -> visitor::Result {
-        match (&left.kind, &right.kind) {
+        match (left.kind, right.kind) {
             // we can't compare with tuples, so we'll convert it to an AND
             (ExpressionKind::Row(left), ExpressionKind::Row(right)) => {
-                self.visit_multiple_tuple_comparison(left.clone(), Values::from(iter::once(right.clone())), true)?;
+                self.visit_multiple_tuple_comparison(left, Values::from(iter::once(right)), true)?;
             }
-            (_left_kind, _right_kind) => {
+            (left_kind, right_kind) => {
+                let (l_alias, r_alias) = (left.alias, right.alias);
+
                 #[cfg(feature = "xml")]
-                let left_xml = left.is_xml_value();
+                let (left_xml, right_xml) = (left_kind.is_xml_value(), right_kind.is_xml_value());
+
+                let mut left = Expression::from(left_kind);
+
+                if let Some(alias) = l_alias {
+                    left = left.alias(alias);
+                }
+
+                let mut right = Expression::from(right_kind);
+
+                if let Some(alias) = r_alias {
+                    right = right.alias(alias);
+                }
 
                 match self {
                     #[cfg(feature = "xml")]
-                    _ if right.is_xml_value() => {
+                    _ if right_xml => {
                         self.surround_with("CAST(", " AS NVARCHAR(MAX))", |x| x.visit_expression(left))?
                     }
                     _ => self.visit_expression(left)?,
```
```

Reply:
Was wondering if there's an easy way to get rid of the clone calls.

Reply:
This is not that "easy", but something to put you going. You can make this even better if you like, kind of noisy my changes...

Reply:
The aliasing could be nicer, but better go to sleep and stop working goddammit...